<a href="https://colab.research.google.com/github/drpetros11111/Tensorflow_Portilia/blob/Numpy/00_NumPy_Arrays.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

___

<a href='http://www.pieriandata.com'><img src='../Pierian_Data_Logo.png'/></a>
___
<center><em>Copyright Pierian Data</em></center>
<center><em>For more information, visit us at <a href='http://www.pieriandata.com'>www.pieriandata.com</a></em></center>

# NumPy

NumPy is a powerful linear algebra library for Python. What makes it so important is that almost all of the libraries in the <a href='https://pydata.org/'>PyData</a> ecosystem (pandas, scipy, scikit-learn, etc.) rely on NumPy as one of their main building blocks. Plus we will use it to generate data for our analysis examples later on!

NumPy is also incredibly fast, as it has bindings to C libraries. For more info on why you would want to use arrays instead of lists, check out this great [StackOverflow post](http://stackoverflow.com/questions/993984/why-numpy-instead-of-python-lists).

We will only learn the basics of NumPy.

## Using NumPy

Once you've installed NumPy you can import it as a library:

In [ ]:
import numpy as np

NumPy has many built-in functions and capabilities. We won't cover them all but instead we will focus on some of the most important aspects of NumPy: vectors, arrays, matrices and number generation. Let's start by discussing arrays.

# NumPy Arrays

NumPy arrays are the main way we will use NumPy throughout the course. NumPy arrays essentially come in two flavors: vectors and matrices. Vectors are strictly 1-dimensional (1D) arrays and matrices are 2D (but you should note a matrix can still have only one row or one column).

Let's begin our introduction by exploring how to create NumPy arrays.

## Creating NumPy Arrays

### From a Python List

We can create an array by directly converting a list or list of lists:

In [1]:
my_list = [1,2,3]
my_list

[1, 2, 3]

In [3]:
import numpy as np

np.array(my_list)

array([1, 2, 3])

In [4]:
my_matrix = [[1,2,3],[4,5,6],[7,8,9]]
my_matrix

[[1, 2, 3], [4, 5, 6], [7, 8, 9]]

In [5]:
np.array(my_matrix)

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

# Np Array
When you use np.array(my_matrix) to transform a matrix into an array, you're converting a matrix-like object (such as a NumPy matrix or a nested list) into a NumPy array. Here's what you achieve by doing this:

--------------------------
#1. Convert to a General NumPy Array
NumPy matrix objects are specifically designed for matrix algebra, but they have some limitations.

For example, a matrix is always 2-dimensional.

NumPy arrays are more flexible and can be multi-dimensional (1D, 2D, 3D, etc.), making them more versatile for numerical computations beyond just matrix operations.

By converting a matrix to an array, you unlock the broader functionality of NumPy arrays, which are used in more contexts than just linear algebra.

-----------------------------------
#2. Access to Element-Wise Operations
With NumPy arrays, operations are typically performed element-wise by default.

This means that if you perform operations like addition, subtraction, multiplication, or division on two arrays, the operations will be applied element by element.

In contrast, with NumPy matrices, * refers to matrix multiplication, and element-wise multiplication requires explicit functions like np.multiply().
For example:

    a = np.array([[1, 2], [3, 4]])
    b = np.array([[2, 2], [2, 2]])
    print(a * b)  # element-wise multiplication

This is different from a NumPy matrix, where * would perform matrix multiplication.

---------------------

#3. More Consistency in NumPy Functions
Many NumPy functions expect arrays rather than matrices as input.

While NumPy supports both arrays and matrices, many high-level functions and libraries (like scipy, pandas, and matplotlib) work more smoothly with arrays.

Converting a matrix to an array ensures compatibility with a wider range of NumPy functions and avoids unexpected behavior.

-------------------------------------
#4. Access to Multi-Dimensional Data Structures
NumPy matrices are restricted to 2 dimensions, while arrays can have more dimensions. Converting to a NumPy array allows you to represent data with higher dimensions, such as tensors (used in deep learning).

For example:

    my_3d_array = np.array([[[1, 2], [3, 4]], [[5, 6], [7, 8]]])
    print(my_3d_array.shape)  # Outputs (2, 2, 2) - a 3D array

----------------------------
#5. Deprecation of the Matrix Class
The NumPy matrix class is discouraged for new code and is considered less flexible than arrays.

The array class is preferred in modern NumPy code due to its greater versatility and generality.

If you're working with matrices, converting them to arrays makes your code more future-proof and easier to integrate with current NumPy practices.

Example:

    import numpy as np
    my_matrix = np.matrix([[1, 2], [3, 4]])
    print(type(my_matrix))  # <class 'numpy.matrix'>

    my_array = np.array(my_matrix)
    print(type(my_array))  # <class 'numpy.ndarray'>

Here, my_matrix is a NumPy matrix, and after the conversion, my_array is a NumPy array (numpy.ndarray).

-------------------------------------

#Summary of Benefits:
Flexibility: Arrays support more than two dimensions.

Element-wise operations: By default, operations are performed element-wise with arrays.

Wider compatibility: NumPy arrays work with more functions and libraries.

Future-proofing: Arrays are preferred in modern code over matrices.

In short, converting a matrix to an array makes the data structure more flexible, consistent, and widely compatible for various operations in NumPy and other libraries.

## Built-in Methods

There are lots of built-in ways to generate arrays.

### arange

Return evenly spaced values within a given interval. [[reference](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.arange.html)]

In [ ]:
np.arange(0,10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [ ]:
np.arange(0,11,2)

array([ 0,  2,  4,  6,  8, 10])

### zeros and ones

Generate arrays of zeros or ones. [[reference](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.zeros.html)]

In [ ]:
np.zeros(3)

array([0., 0., 0.])

In [ ]:
np.zeros((5,5))

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

In [ ]:
np.ones(3)

array([1., 1., 1.])

In [6]:
np.ones((3,3))

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

### linspace
Return evenly spaced numbers over a specified interval. [[reference](https://www.numpy.org/devdocs/reference/generated/numpy.linspace.html)]

In [7]:
np.linspace(0,10,3)

array([ 0.,  5., 10.])

In [8]:
np.linspace(0,5,20)

array([0.        , 0.26315789, 0.52631579, 0.78947368, 1.05263158,
       1.31578947, 1.57894737, 1.84210526, 2.10526316, 2.36842105,
       2.63157895, 2.89473684, 3.15789474, 3.42105263, 3.68421053,
       3.94736842, 4.21052632, 4.47368421, 4.73684211, 5.        ])

# Linespace
The np.linspace(0, 5, 20) function from the NumPy library generates an array of evenly spaced numbers over a specified range. Here's a detailed explanation of its components:

-------------------------
Syntax:
    np.linspace(start, stop, num)

Where:

start: The starting value of the range (in this case, 0).

stop: The ending value of the range (in this case, 5).

num: The number of evenly spaced samples to generate (in this case, 20).

Explanation of np.linspace(0, 5, 20):

Range:

The function generates numbers from 0 to 5 (inclusive of both endpoints).
Number of Samples:

20 means that it will create 20 evenly spaced numbers between 0 and 5.
Evenly Spaced:

The interval between each number is calculated so that the numbers are evenly distributed between the starting value 0 and the ending value 5.

Example:

    import numpy as np
    result = np.linspace(0, 5, 20)
    print(result)

This produces an array:

[0.         0.26315789 0.52631579 0.78947368 1.05263158 1.31578947
 1.57894737 1.84210526 2.10526316 2.36842105 2.63157895 2.89473684
 3.15789474 3.42105263 3.68421053 3.94736842 4.21052632 4.47368421
 4.73684211 5.        ]

As you can see, the numbers start at 0, end at 5, and there are exactly 20 values between them, evenly spaced.

-----------------------------
#Key Points:
Inclusive of endpoints: Both 0 (start) and 5 (stop) are included in the result.

Even spacing: The values are evenly distributed, meaning the distance between consecutive elements is constant.

Useful for plotting: np.linspace() is often used to generate values for the x-axis in plotting functions (e.g., creating a smooth curve).

Formula for the spacing:
The spacing between consecutive elements is:

$$spacing
=
𝑠
𝑡
𝑜
𝑝
−
𝑠
𝑡
𝑎
𝑟
𝑡
𝑛
𝑢
𝑚
−
1
spacing=
num−1
stop−start$$
​
So each consecutive value is separated by approximately 0.2632.

----------------------
#Summary
np.linspace(0, 5, 20) generates 20 equally spaced numbers from 0 to 5, including both 0 and 5.

It's commonly used in numerical computations and plotting where evenly spaced values over a specific range are needed.

<font color=green>Note that `.linspace()` *includes* the stop value. To obtain an array of common fractions, increase the number of items:</font>

In [ ]:
np.linspace(0,5,21)

array([0.  , 0.25, 0.5 , 0.75, 1.  , 1.25, 1.5 , 1.75, 2.  , 2.25, 2.5 ,
       2.75, 3.  , 3.25, 3.5 , 3.75, 4.  , 4.25, 4.5 , 4.75, 5.  ])

### eye

Creates an identity matrix [[reference](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.eye.html)]

In [ ]:
np.eye(4)

array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])

## Random
Numpy also has lots of ways to create random number arrays:

### rand
Creates an array of the given shape and populates it with random samples from a uniform distribution over ``[0, 1)``. [[reference](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.random.rand.html)]

In [9]:
np.random.rand(2)

array([0.41891664, 0.2684813 ])

In [10]:
np.random.rand(5,5)

array([[0.6195457 , 0.86845742, 0.06145989, 0.54057133, 0.42789311],
       [0.40057192, 0.73054891, 0.38779001, 0.61517892, 0.99134199],
       [0.46921519, 0.26025239, 0.20410361, 0.74921543, 0.14842917],
       [0.07579683, 0.82750188, 0.8459314 , 0.44299275, 0.70651644],
       [0.7228766 , 0.33269093, 0.76077668, 0.67342665, 0.84068824]])

### randn

Returns a sample (or samples) from the "standard normal" distribution [σ = 1]. Unlike **rand** which is uniform, values closer to zero are more likely to appear. [[reference](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.random.randn.html)]

In [11]:
np.random.randn(2)

array([ 2.19051732, -0.57149105])

In [ ]:
np.random.randn(5,5)

array([[-0.45241033,  1.07491082,  1.95698188,  0.40660223, -1.50445807],
       [ 0.31434506, -2.16912609, -0.51237235,  0.78663583, -0.61824678],
       [-0.17569928, -2.39139828,  0.30905559,  0.1616695 ,  0.33783857],
       [-0.2206597 , -0.05768918,  0.74882883, -1.01241629, -1.81729966],
       [-0.74891671,  0.88934796,  1.32275912, -0.71605188,  0.0450718 ]])

### randint
Returns random integers from `low` (inclusive) to `high` (exclusive).  [[reference](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.random.randint.html)]

In [12]:
np.random.randint(1,100)

19

In [13]:
np.random.randint(1,100,10)

array([ 1, 28, 48, 11, 62, 92,  8,  9, 88, 55])

### seed
Can be used to set the random state, so that the same "random" results can be reproduced. [[reference](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.random.seed.html)]

In [14]:
np.random.seed(42)
np.random.rand(4)

array([0.37454012, 0.95071431, 0.73199394, 0.59865848])

In [15]:
np.random.seed(42)
np.random.rand(4)

array([0.37454012, 0.95071431, 0.73199394, 0.59865848])

## Array Attributes and Methods

Let's discuss some useful attributes and methods for an array:

In [16]:
arr = np.arange(25)
ranarr = np.random.randint(0,50,10)

In [17]:
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24])

In [18]:
ranarr

array([38, 18, 22, 10, 10, 23, 35, 39, 23,  2])

## Reshape
Returns an array containing the same data with a new shape. [[reference](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.reshape.html)]

In [19]:
arr.reshape(5,5)

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24]])

### max, min, argmax, argmin

These are useful methods for finding max or min values. Or to find their index locations using argmin or argmax

In [20]:
ranarr

array([38, 18, 22, 10, 10, 23, 35, 39, 23,  2])

In [21]:
ranarr.max()

39

In [22]:
ranarr.argmax()

7


The function ranarr.argmax() is a NumPy method that returns the index of the maximum value in the array ranarr.

In [ ]:
ranarr.min()

2

In [ ]:
ranarr.argmin()

9

## Shape

Shape is an attribute that arrays have (not a method):  [[reference](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.ndarray.shape.html)]

In [ ]:
# Vector
arr.shape

(25,)

In [ ]:
# Notice the two sets of brackets
arr.reshape(1,25)

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15,
        16, 17, 18, 19, 20, 21, 22, 23, 24]])

In [ ]:
arr.reshape(1,25).shape

(1, 25)

In [ ]:
arr.reshape(25,1)

array([[ 0],
       [ 1],
       [ 2],
       [ 3],
       [ 4],
       [ 5],
       [ 6],
       [ 7],
       [ 8],
       [ 9],
       [10],
       [11],
       [12],
       [13],
       [14],
       [15],
       [16],
       [17],
       [18],
       [19],
       [20],
       [21],
       [22],
       [23],
       [24]])

In [ ]:
arr.reshape(25,1).shape

(25, 1)

### dtype

You can also grab the data type of the object in the array: [[reference](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.ndarray.dtype.html)]

In [ ]:
arr.dtype

dtype('int32')

In [ ]:
arr2 = np.array([1.2, 3.4, 5.6])
arr2.dtype

dtype('float64')

# Great Job!